File: completed/1717.java
import java.util.*;

class Solution {
    public int maximumGain(String s, int x, int y) {
        Stack<Character> letters = new Stack<>();
        int total = 0;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if ((c != 'a' && c != 'b') || letters.empty()) {
                letters.push(c);
                continue;
            }

            if (x > y) {
                if (letters.peek() == 'a' && c == 'b') {
                    letters.pop();
                    total += x;
                } else {
                    letters.push(c);
                }
            } else {
                if (letters.peek() == 'b' && c == 'a') {
                    letters.pop();
                    total += y;
                } else {
                    letters.push(c);
                }
            }
        }

        Stack<Character> previous = new Stack<>();
        while (!letters.empty()) {
            char c = letters.pop();
            if ((c != 'a' && c != 'b') || previous.empty()) {
                previous.push(c);
                continue;
            }

            if (x > y) {
                if (previous.peek() == 'a' && c == 'b') {
                    previous.pop();
                    total += y;
                } else {
                    previous.push(c);
                } 
            } else {
                if (previous.peek() == 'b' && c == 'a') {
                    previous.pop();
                    total += x;
                } else {
                    previous.push(c);
                }
            }
        }

        return total;
    }
}

File: completed/1717.cpp
class Solution {
public:
    int maximumGain(string s, int x, int y) {
        int score = 0;

        stack<char> forward;
        for (int i = 0; i < s.length(); i++) {
            if ((s[i] != 'a' && s[i] != 'b') || forward.empty()) {
                forward.push(s[i]);
                continue;
            }

            if (x > y) {
                if (s[i] == 'b' && forward.top() == 'a') {
                    forward.pop();
                    score += x;
                } else {
                    forward.push(s[i]);
                }
            } else {
                if (s[i] == 'a' && forward.top() == 'b') {
                    forward.pop();
                    score += y;
                } else {
                    forward.push(s[i]);
                }
            }
        }

        stack<char> backward;
        while (!forward.empty()) {
            char c = forward.top();
            forward.pop();

            if ((c != 'a' && c != 'b') || backward.empty()) {
                backward.push(c);
                continue;
            }

            if (x > y) {
                if (c == 'b' && backward.top() == 'a') {
                    backward.pop();
                    score += y;
                } else {
                    backward.push(c);
                }
            } else {
                if (c == 'a' && backward.top() == 'b') {
                    backward.pop();
                    score += x;
                } else {
                    backward.push(c);
                }
            }
        }
        return score;
    }
};
